src/libstd/sys/unix/thread.rs

```diff
@@ -71,7 +71,7 @@ impl Thread {
         };
 
         extern fn thread_start(main: *mut libc::c_void) -> *mut libc::c_void {
-            unsafe { start_thread(main); }
+            unsafe { start_thread(main, pthread_self() as u32); }
```

Comment:
That could be an option. Using `pthread_equals` (and `HANDLE` equality of some sort on Windows) could be another.